### PR TITLE
library: first implementation of the XBee Android Library

### PR DESCRIPTION
**Pull request by diescalo**
- The project has just one module for the library: xbee_android_library
- Moved here all the Android stuff from the XBee Java Library project
  (https://github.com/digidotcom/XBeeJavaLibrary)
- dependencies of the library should be revisited before releasing it.

Signed-off-by: Diego Escalona <diego.escalona@digi.com>

### CODE REVIEW

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/XBeeDevice.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    /**
+     * Class constructor. Instantiates a new {@code XBeeDevice} object for
+     * Android with the given parameters.
+     *
+     * <p>This constructor uses the Android Bluetooth Low Energy API to
+     * communicate with the devices.</p>
+     *
+     * <p>The Bluetooth password must be provided before calling the
+     * {@link #open()} method, either through this constructor or the
+     * {@link #setBluetoothPassword(String)} method.</p>
```
Comment (rubenmoral) [inline xbee_android_library/src/main/java/com/digi/xbee/api/android/XBeeDevice.java]: I would change the visibility of the `setBluetoothPassword(String)` method to `protected` in the Java library and make it public in these subclasses.

Comment (diescalo) [reply xbee_android_library/src/main/java/com/digi/xbee/api/android/XBeeDevice.java]: Will do.

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		write(buffer, 0, buffer.length);
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see java.io.OutputStream#write(byte[], int, int)
+	 */
+	@Override
+	public void write(byte[] buffer, int offset, int count) throws IOException {
+		final byte[] finalData = new byte[count + offset];
```
Comment (mikewadsten) [inline xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java]: Do we really want to launch a new thread for every write call? Seems more logical to me to have a long-running thread onto which we schedule the writes.

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidXBeeInterface.java [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+	 * the FT232BM, FT2232C and FT232LR chips.
+	 * 
+	 * @param baud Desired baud rate. 
+	 * @param base The base clock.
+	 * 
+	 * @return The calculated divisor.
+	 */
+	private int calculateBaudRate(int baud, int base) {
+		int divisor;
+		divisor = (base / 16 / baud) | (((base / 2 / baud) & 4) != 0 ? 0x4000 // 0.5
```
Comment (mikewadsten) [inline xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidXBeeInterface.java]: Could this be rewritten to not depend on chained ternary expressions?

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java (hunk 2 above)
Comment (diescalo) [reply xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java]: To be totally honest I didn't took a deep look to the code as I was just moving classes from one lib to another. Anyway you're right, a single thread taking data from a queue is a much better approach. I'll change it =)

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidXBeeInterface.java (hunk 3 above)
Comment (diescalo) [reply xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidXBeeInterface.java]: I'll change the method to:
```java
	private int calculateBaudRate(int baud, int base) {
		int divisor = (base / 16 / baud);
		if (((base / 2 / baud) & 4) != 0)
			divisor |= 0x4000; // 0.5
		else if (((base / 2 / baud) & 2) != 0)
			divisor |= 0x8000; // 0.25
		else if (((base / 2 / baud) & 1) != 0)
			divisor |= 0xC000; // 0.125
		else
			divisor |= 0;

		return divisor;
	}
```

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java (hunk 2 above)
Comment (diescalo) [reply xbee_android_library/src/main/java/com/digi/xbee/api/android/connection/android/AndroidUSBOutputStream.java]: This is the new implementation so far:
```java
private LinkedBlockingQueue<byte[]> writeQueue;

public AndroidUSBOutputStream(UsbEndpoint writeEndpoint, UsbDeviceConnection connection) {
	[...]

	writeQueue = new LinkedBlockingQueue<>(512);
	DataWriter dataWriter = new DataWriter();
	dataWriter.start();
}

public void write(byte[] buffer, int offset, int count) {
	final byte[] finalData = new byte[count + offset];
	System.arraycopy(buffer, offset, finalData, 0, count);
	try {
		writeQueue.add(finalData);
	} catch (IllegalStateException  e) {
		logger.error("Could not add data, write queue is full: " + e.getMessage(), e);
	}
}

/**
 * Internal class used to write data coming from a queue.
 */
class DataWriter extends Thread {
	@Override
	public void run() {
		while (streamOpen) {
			try {
				byte[] dataToWrite = writeQueue.poll(100, TimeUnit.MILLISECONDS);
				if (dataToWrite == null)
					continue;
				usbConnection.bulkTransfer(sendEndPoint, dataToWrite, dataToWrite.length, WRITE_TIMEOUT);
				logger.debug("Message sent: " + HexUtils.byteArrayToHexString(dataToWrite));
			} catch (InterruptedException e) {
				logger.error("Interrupted while getting data from the write queue: " + e.getMessage(), e);
			}
		}
	}
}

@Override
public void close() throws IOException {
	// Stop the data writer.
	streamOpen = false;
	super.close();
}
```

File: xbee_android_library/src/main/java/com/digi/xbee/api/android/XBeeDevice.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    @Override
+    public void sendSerialData(byte[] data) throws XBeeException {
+        super.sendSerialData(data);
+    }
+
+    @Override
+    public void addSerialDataListener(ISerialDataReceiveListener listener) {
+        super.addSerialDataListener(listener);
+    }
+}
```
Comment (rubenmoral) [inline xbee_android_library/src/main/java/com/digi/xbee/api/android/XBeeDevice.java]: Missing `removeSerialDataListener` methods.

**Pull request merged**